### ISSUE TITLE
Where can algebraic universes be used?

### ISSUE THREAD

**Issue by jashug** [coq/coq#10423]
TL;DR: Currently, there are lots of anomalies with algebraic universes, but allowing depth-1 conclusion-algebraic types (but not algebraic universes in terms of a universe) might be a reasonable restriction.

Currently, there is syntax to write `Type@{i+1}`, or `Type@{max(i,j+1)}`, though not `Type@{i+2}`.
Writing these types sometimes results in the anomaly [`Unable to handle arbitrary u+k <= v constraints.`](https://github.com/coq/coq/blob/d886dff0857702fc4524779980ee6b7e9688c1d4/kernel/univ.ml#L690) (see ex. #7420, #6678)

It would be good to add restrictions on the use of algebraic universes and error checking to avoid these anomalies. In this issue I would like to argue that forbidding algebraic universes everywhere is too strong of a restriction.

Algebraic types do arise, for example as the type of `U` in `Definition U@{i} := Type@{i}` or with inductive types. Having syntax to annotate these definitions with, as `Definition U@{i} : Type@{i+1} := Type@{i}.` is useful at the very least for human understanding.

Working off my understanding of [1], I would like to think that (depth-1) algebraic universes do no harm when they are in type position, that is, I hope allowing conclusion-algebraic types (but not terms) should be manageable. Particularly if we restrict to depth-1 conclusion-algebraic types, this would imply that the annotation `A` in (fun x : A => y : B) cannot be algebraic, but that B could be conclusion-algebraic, and that for `let x : A := t in y`, A could be conclusion-algebraic.

The other place I am aware of types appearing in terms (as opposed to terms of a universe) is in the return clause of a match statement (perhaps also the return type of a fix?), and this is related to #10422. Currently, depth-0 algebraic types work, but depth-1 types do not.

References:
[1] Herbelin, Hugo @herbelin [Type inference with algebraic universes in the
Calculus of Inductive Constructions](https://pdfs.semanticscholar.org/ab75/e0b02aa47fdd783e8f5ba88d2c9bcb860db5.pdf)

**Comment by SkySkimmer** [comment]
What is depth here?

See also https://github.com/coq/coq/pull/7490

**Comment by jashug** [comment]
depth (a term from [1]) is the longest chain of successor operations in the algebraic, so `max(i,j)` is depth 0 while `i+1` and `max(i,j+1)` are depth 1, while `i+2` would be depth 2.
conclusion-algebraic means either non-algebraic, `Type@{e}` with e algebraic, or `forall x : X, T` where X is non-algebraic and T is conclusion algebraic.

Particularly, Corollary 1 from [1] is interesting, saying that if `If Γ and t are non algebraic and Γ |- t : T then T is depth-1 conclusion-algebraic` (plus a condition about constraints)

**Comment by jashug** [comment]
Concretely, my suggestion is that algebraic universes should be allowed exactly in the conclusion of types.

Comparing that to #7490, I would allow the type of let-ins to be conclusion-algebraic, to handle `let U := Type@{i} in y`, since uses of `U` only need to check `i+1 <= j`, which is doable.

Checking that `t : T` with `T` conclusion algebraic is hard since it involves comparing two algebraic universes (see #10425), but my opinion is that requiring the equation to hold without adding new constraints in this situation is a good option (we already do this, but give an anomaly rather than a user-error when it doesn't hold). In particular, if `T` is exactly the inferred type, then we are comparing `e <= e`, which always holds.
